lighthouse-core/scripts/pptr-run-devtools.js

```diff
@@ -152,122 +176,146 @@ const startLighthouse = `
       }
     }
   }
+}
+
+async function runLighthouse() {
+  // @ts-expect-error global
+  const panel = UI.panels.lighthouse || UI.panels.audits;
+
+  /** @type {Promise<{lhr: LH.Result, artifacts: LH.Artifacts}>} */
+  const resultPromise = new Promise((resolve, reject) => {
+    const methodName = panel.__proto__.buildReportUI ?
+      'buildReportUI' : '_buildReportUI';
+    addSniffer(
+      panel.__proto__,
+      methodName,
+      // @ts-expect-error implicit any
+      (lhr, artifacts) => resolve({lhr, artifacts})
+    );
+
+    addSniffer(
+      panel.statusView.__proto__,
+      'renderBugReport',
+      // @ts-expect-error implicit any
+      (err) => reject(err)
+    );
+  });
 
+  const button = panel.contentElement.querySelector('button');
   button.click();
-})()
-`;
+
+  return resultPromise;
+}
+/* eslint-enable */
 
 /**
- * @param {string} url
+ * @param {puppeteer.Browser} browser
+ * @param {puppeteer.CDPSession} inspectorSession
+ * @param {LH.Config.Json} config
  */
-function isValidUrl(url) {
-  try {
-    new URL(url);
-    return true;
-  } catch {
-    return false;
+async function installCustomConfig(browser, inspectorSession, config) {
+  // Screen emulation is handled by DevTools, so we should avoid adding our own.
+  if (config.settings?.screenEmulation) {
+    throw new Error('Configs that modify device emulation are unsupported in DevTools');
   }
+  if (!config.settings) config.settings = {};
+  config.settings.screenEmulation = {disabled: true};
+
+  const workerTarget = await browser.waitForTarget(t => t.url().includes('lighthouse_worker'));
+
+  // Ensure the inspector is paused once the worker is initialized so Lighthouse doesn't start prematurely.
+  await Promise.all([
+    inspectorSession.send('Runtime.enable'),
+    inspectorSession.send('Debugger.enable'),
+  ]);
+  await inspectorSession.send('Debugger.pause');
+
+  const workerSession = await workerTarget.createCDPSession();
+  await Promise.all([
+    workerSession.send('Runtime.enable'),
+    workerSession.send('Debugger.enable'),
+    workerSession.send('Runtime.runIfWaitingForDebugger'),
+    new Promise(resolve => {
+      workerSession.once('Debugger.scriptParsed', resolve);
+    }),
+  ]);
+  await evaluateInSession(workerSession,
+    `self.createConfig = () => (${JSON.stringify(config)})`
+  );
+
+  await inspectorSession.send('Debugger.resume');
+}
+
+/**
+ * @param {puppeteer.CDPSession} inspectorSession
+ * @param {string[]} logs
+ */
+async function installConsoleListener(inspectorSession, logs) {
```

Comment:
We were previously using the stdout and stderr of this script on the CLI as our log for the run. Listening to console messages on the inspector target will allow us to debug with `console.log` in the Lighthouse bundle.